remove setStorage with null StorageImpl support.

gchanan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #33823 Get rid of newWithStorage2d calls.
* #33815 Kill some unused (TH)Storage-based APIs.
* **#33735 remove setStorage with null StorageImpl support.**
* #33734 Remove some dead THStorage related code.
* #33695 Kill dead scalar_check.

This apparently used to create a new storage, but I couldn't find anywhere in the code where this actually happens.

Changing it to an assert to see what happens.

Differential Revision: [D20084029](https://our.internmc.facebook.com/intern/diff/D20084029)

gchanan committed:
remove setStorage with null StorageImpl support.

This apparently used to create a new storage, but I couldn't find anywhere in the code where this actually happens.

Changing it to an assert to see what happens.

gchanan committed:
Update on "remove setStorage with null StorageImpl support."


This apparently used to create a new storage, but I couldn't find anywhere in the code where this actually happens.

Changing it to an assert to see what happens.

Differential Revision: [D20084029](https://our.internmc.facebook.com/intern/diff/D20084029)

bhosmer commented on aten/src/TH/THTensor.cpp:
```diff
@@ -48,14 +48,13 @@ void THTensor_setStorageNd(THTensor *self, THStorage *storage, ptrdiff_t storage
     if(storage)
     {
       c10::raw::intrusive_ptr::incref(storage);
       THTensor_stealAndSetStoragePtr(self, storage);
```
Probably a pointless question, but - if this is a real steal, like a move, does the ref count get decremented back down in the call? Or is this just sharing the storage?

gchanan replied:
which refcount?

There are two storages here, the tensor's old storage and the tensor's new storage.

For the old storage the refcount is decreased because the tensor no longer points to the old storage at the end/
For the new storage the refcount is increased, i.e. the steal is on the incref.

All of this seem pretty pointless:
1) we don't need a steal API anymore as we don't pass in a tensor that needs to be stolen
2) (see other comment) we don't actually need two storages here.

bhosmer replied:
Yeah, thanks for the background! It would definitely be sweet to get rid of the vestigial degrees of freedom. 

Just to clarify, I was thinking of the `incref(storage)` - figuring that if the storage was literally getting stolen (i.e. destructively acquired, move style) then the abstract decref/incref pair would net out to no change. My guess was that "steal" here really means "alias" or maybe that there was some deeper subtlety, just wanted to drill down. 

...but really, happy not to dedicate neural real estate on an API that's on its way out 😬

gchanan replied:
I believe "steal" here refers to stealing the refcount -- I'm not sure if that fits your definition of "destructively acquired, move style."  It does if we are talking about ref counted pointers, although that's not obvious here.

bhosmer commented on aten/src/TH/generic/THTensor.cpp:
```diff
@@ -74,11 +74,12 @@ THTensor *THTensor_(newWithStorage)(THStorage *storage, ptrdiff_t storageOffset,
     at::DispatchKey::CPUTensorId
   ).release();
-  THTensor_(setStorageNd)(self, storage, storageOffset, sizes.size(),
+  THTensor_(setStorageNd)(self, storage != nullptr ? storage : new_storage, storageOffset, sizes.size(),
```
I know it's not part of the logic change in the PR but I'm educationally curious what's happening here, like why we create a new `THStorage` even if a non-null one is passed. 

Is it that `THStorage` is just a wrapper and so `new_storage` gets the expensive part of `storage` plugged into it if `storage` is non-null?

gchanan replied:
I think it's historical -- storages and tensors used to be more disconnected, so it was possible to e.g. have a tensor without a storage (or to swap the type of a tensor by just replacing its storage).

This kind of forced storages to be first class and had a bunch of corner cases -- i.e. do I need to handle the case where a tensor doesn't have a storage in each of my functions?

Later we decided to disallow those cases, which made the APIs above really convoluted (I need to create a storage because I always need a storage, but I'm just going to immediately wipe out that storage anyway).

So I really think we should just get rid of all these storage-focused APIs.  I _might_ do that in a follow on PR as there aren't many cases left.

bhosmer reviewed (approved):
Tangential questions inline but we can follow up offline if easier

gchanan commented:
@pytorchbot retest this please.

gchanan committed:
Update on "remove setStorage with null StorageImpl support."


This apparently used to create a new storage, but I couldn't find anywhere in the code where this actually happens.

Changing it to an assert to see what happens.

Differential Revision: [D20084029](https://our.internmc.facebook.com/intern/diff/D20084029)
